C++ implementation of the Stacked Faith algorithm

jepasan opened the pull request:
This PR implements a fast C++ algorithm (via Rcpp) for calculating Faith's index, based on the C++ code from the sources in #522. For simplicity, I ended up extracting the Faith-related code and condensing it into just a few source files, and doing some amount of rewriting to simplify the code and integrate it more smoothly with R.

The algorithm can be toggled with the parameter `fast.faith` to `addAlpha()` (`TRUE` by default). Based on my own testing, even large datasets have their indices calculated pretty much instantly.

```
> tse <- microbiomeDataSets::GrieneisenTSData()
> system.time(x <- addAlpha(tse, index="faith", fast.faith=TRUE))
   user  system elapsed 
   0.47    0.03    0.50 
> system.time(x2 <- addAlpha(tse, index="faith", fast.faith=FALSE))
   user  system elapsed 
  35.93    0.47   36.92 
> sum(abs(colData(x)$faith - colData(x2)$faith) > 0.00000001)
[1] 7
```

As you can see in the example, the output of the new algorithm may sometimes differ somewhat from the old one. This is because the new algorithm always includes lengths all the way down to the root in its calculations, while the old algorithm seems to drop them sometimes during the process of pruning the tree. In concrete terms, the new algoritm matches the output of `picante::pd()` with `include.root=TRUE`, while the old algorithm matches `include.root=FALSE`. For most datasets I checked this did not lead to significant differences, but for e.g. `SprockettTHData` from `microbiomeDataSets` there is a consistent difference of ~1.17 for basically every sample. I don't know how significant this is going to be in practice, however.

Another thing to note is that the original C++ code is licensed under the 3-clause BSD license - this is compatible with mia's licensing, but I'm not sure what, if anything, needs to be done to properly indicate this in the package.

jepasan committed:
Add C++ code from unifrac's implementation

jepasan committed:
Copy faith-related code into separate files

jepasan committed:
Fix code crashing due to out-of-bounds vector access

jepasan committed:
C++ code functional for some datasets

The C++ code works for rooted trees, but not unrooted ones. Possibly a bug in my implementation, more testing needed.

jepasan committed:
Clean up unnecessary files

jepasan committed:
Simplify source files and code

jepasan committed:
Modify folder structure to fit Rcpp guidelines

jepasan committed:
Integrate Rcpp and C++ code into the package

Some datasets still produce divergent values. Likely a bug in my implementation.

jepasan committed:
Fix the bug that was causing divergent results

Bringing the assay into C++ was using rowTree tip labels for the observation ids, causing nonsense results when they were in different order from the actual rownames.

Also added a check for cladewise tree ordering.

jepasan committed:
Miscellaneous cleaning up

jepasan committed:
Update documentation

jepasan committed:
Add all object files to .gitignore

jepasan committed:
Rename header files to get rid of warnings

TuomasBorman reviewed (changes requested):
Looks very good!

Details section of `addAlpha` should mention clearly the origin of the implementation. 

You can put the info here:

```
#'  \item 'faith': Faith's phylogenetic alpha diversity index measures how
#' long the taxonomic distance is between taxa that are present in the sample.
#' Larger values represent higher diversity. Using this index requires
#' rowTree. (Faith 1992)
```

antagomir commented:
Awesome.

BSD3 is the same as BSD2, but additionally prohibiting the names of the authors from being used to endorse or promote the software. If you could mention the source (url) where you took parts of the code from this should be sufficient. The function roxygen documentation @details section would be a suitable place I think.

Imo the differences are acceptable as long as the @details section in the roxygen documentation would mention them clearly. The function works correctly and the differences are justified and transparent.
 
Some more comments in the code PR.

antagomir commented on R/estimateDiversity.R:
```diff
@@ -115,19 +115,34 @@
+NULL
+
+#' @importFrom ape reorder.phylo
+.calc_faith <- function(mat, tree, only.tips = FALSE, fast.faith = TRUE, ...){
```
I am not sure if we want to keep the old Faith index at all - could we just remove it?

antagomir replied:
@TuomasBorman

TuomasBorman replied:
I think we can remove it, there is no need for it anymore, I guess.

Removing the old implementation also simplifies things

antagomir commented on src/faith_R.cpp:
```diff
@@ -0,0 +1,70 @@
+ * This functions makes several assumptions about treeSE:
+ * 
+ *  - It must contain a non-empty counts assay and a non-empty RowTree
+ *  - The RowTree must be sorted in cladewise order
+ *  - The RowTree must be rooted - Unrooted trees can be passed without error, but don't produce correct results
```
RowTree -> rowTree

antagomir replied:
shall we throw a warning with an unrooted tree?

antagomir commented on src/faith_R.cpp:
```diff
@@ -0,0 +1,70 @@
+ * treeSE <const Rcpp::S4 &> an R TreeSummarizedExperiment object.
+ * faith <Rcpp::NumericVector> the resulting vector of computed Faith PD values
+ *
+ * This functions makes several assumptions about treeSE:
```
treeSE -> TreeSE usually

antagomir commented:
I suggest to 

- remove the old and slow Faith implementation entirely (or let's think why we should keep it)
- update the documentation accodingly
- add unit tests that show that this yields comparable results with some external implementation

jepasan committed:
Remove old Faith's PD implementation

jepasan committed:
Update documentation

antagomir commented:
Hi @jepasan - could you implement the (hopefully small) additional suggestions, then we could merge?

antagomir committed:
Merge branch 'devel' into fastfaith

jepasan committed:
Merge branch 'fastfaith' of https://github.com/jepasan/mia into fastfaith

jepasan commented:
I've removed the old implementation and updated the documentation. After doing some checking, it seems like the code works fine on both rooted and unrooted trees (it still matches `picante::pd()` in the latter case, just with `include.root=FALSE`). I think the differences were caused by another bug that I fixed, so it shouldn't be a problem anymore.
Regarding unit tests, the existing tests already checks against results from `picante`, which as far as I can tell are still the same, so I'm not sure there's any reason to change them.

jepasan committed:
Make C++ documentation internal

antagomir committed:
Merge branch 'devel' into fastfaith

antagomir committed:
Merge branch 'devel' into fastfaith

antagomir commented on R/RcppExports.R:
```diff
@@ -0,0 +1,39 @@
+#' This function makes several assumptions about the contents of
+#' \code{assay} and \code{rowTree}, namely that:
+#' \itemize{
+#'  \item \code{assay} and \code{rowTree} are both non-empty.
+#'  \item \code{assay} has row and column names.
+#'  \item \code{rowTree}'s nodes are arranged in cladewise order.
+#' }
+#' These checks should all be handled in the surrounding R code.
```
Does the function throw an informative warning or error if these are not met? This could be mentioned in the documentation.

TuomasBorman replied:
1. This is already checked earlier
2. Rownames are added earlier. It should be checked that the function works even when colnames() are NULL. It might be good to check with all the indices.
3. Cladewise sorting does not affect the tree or results, it just sorts the edges

-->  warning might not be needed as these are just internal functionality that do not affect the results

antagomir commented on R/RcppExports.R:
```diff
@@ -0,0 +1,39 @@
+#' by \code{picante::pd()} with the parameter \code{include.root=TRUE}.
+#' 
+#' The C++ code was adapted from an implementation by the Unifrac team (see
+#' \url{https://genome.cshlp.org/content/31/11/2131} or
+#' \url{https://github.com/biocore/unifrac}), which is licensed under the BSD
```
I suggest to add instead a bibliographic information in the @references section and then cite here with (Lastname, year) instead. That give better recognition for the original authors.

Or even better, set up references with this
https://cran.r-project.org/web/packages/Rdpack/vignettes/Inserting_bibtex_references.pdf

antagomir replied:
The references already are in @references section so just update the citing style here.

antagomir reviewed (approved):
Just some small remaining comments attached. Very good.

TuomasBorman commented on R/estimateDiversity.R:
```diff
@@ -115,73 +115,31 @@
-    return(faiths)
+    
+    # The tree must be in cladewise order for the algorithm to work correctly
+    temp <- reorder.phylo(tree, "cladewise")
```
Store back to tree to save memory

TuomasBorman commented on R/RcppExports.R:
```diff
@@ -0,0 +1,39 @@
+#' @export
+.faith_cpp <- function(assay, rowTree) {
+    .Call('_mia_faith_cpp', PACKAGE = 'mia', assay, rowTree)
+}
```
This function should not be exported. It should not be visible for user

TuomasBorman commented on src/faith_R.cpp:
```diff
@@ -0,0 +1,95 @@
+    su::PropMap propmap(table.n_samples);
+    
+    uint32_t node;
+    std::vector<double> node_proportions;
+    double length;
+    
+    std::vector<double> results = std::vector<double>(table.n_samples, 0.0);
+    
+    // for node in postorderselect
+    const unsigned int max_k = (tree_sheared.nparens>1) ? ((tree_sheared.nparens / 2) - 1) : 0;
+    for(unsigned int k = 0; k < max_k; k++) {
+        node = tree_sheared.postorderselect(k);
+        
+        // get branch length
+        length = tree_sheared.lengths[node];
+        
+        // get node proportions and set intermediate scores
+        node_proportions = set_proportions(tree_sheared, node, table, propmap, false);
+        
+        for (unsigned int sample = 0; sample < table.n_samples; sample++){
+            // calculate contribution of node to score
```
Can you still polish the code formatting:

1. All indentations should be multiplication of 4 spaces
2. The line should not exceed 80 characters
3. Check that ifs, for loops, and while loops use same styling (currently this seems not to be harmonized)

antagomir replied:
I think those issues should come up in the automated checks. Have you also run these? like

~/bin/R-x.y.z/bin/R CMD build ../../ #--resave-data #--no-examples  --no-build-vignettes 
~/bin/R-x.y.z/bin/R CMD check mia_a.b.c.tar.gz #--no-build-vignettes --no-examples
~/bin/R-x.y.z/bin/R CMD BiocCheck mia_a.b.c.tar.gz
~/bin/R-x.y.z/bin/R CMD INSTALL mia_a.b.c.tar.gz

jepasan committed:
Remove export keyword

jepasan committed:
Update reference style

jepasan committed:
Merge branch 'fastfaith' of https://github.com/jepasan/mia into fastfaith

antagomir committed:
Merge branch 'devel' into fastfaith

jepasan committed:
Clean up C++ code

jepasan committed:
Save re-sorted edges directly to tree

jepasan committed:
Modify C++ code to work without colnames

jepasan committed:
Merge branch 'fastfaith' of https://github.com/jepasan/mia into fastfaith

jepasan commented:
The code should now work whether or not the assay has colnames. I did also try running the automated checks, but as far as I can tell they don't say anything about the C++ code. Still, I've tried to clean it up as best I could.

antagomir commented:
Seems good to me! @TuomasBorman could confirm?

TuomasBorman committed:
Update docs

TuomasBorman reviewed (approved):
You are correct that BiocCheck does check only formatiing of R files, but good to keep the code format harmonized anyways

I removed roxygen documentation of the c++ function as we do not want that it is visible for user.

Looks good! Can you put your name to DESCRIPTION file? After that, this is ready to be merged

jepasan committed:
Add name to DESCRIPTION

jepasan committed:
Merge branch 'fastfaith' of https://github.com/jepasan/mia into fastfaith
